src/combinatorics/simple_graph/degree_sum.lean

```diff
@@ -0,0 +1,244 @@
+/-
+Copyright (c) 2020 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Kyle Miller.
+-/
+import combinatorics.simple_graph.basic
+import algebra.big_operators.basic
+import data.nat.parity
+import data.zmod.basic
+import tactic.omega
+/-!
+# Degree-sum formula and handshaking lemma
+
+The degree-sum formula is that the sum of the degrees of a finite
+graph is equal to twice the number of edges.  The handshaking lemma is
+a corollary, which is that the number of odd-degree vertices is even.
+
+## Main definitions
+
+- A `dart` is a directed edge, consisting of an ordered pair of adjacent vertices,
+  thought of as being a directed edge.
+- `simple_graph.sum_degrees_eq_twice_card_edges` is the degree-sum formula.
+- `simple_graph.even_card_odd_degree_vertices` is the handshaking lemma.
+- `simple_graph.odd_card_odd_degree_vertices_ne` is that the number of odd-degree
+  vertices different from a given odd-degree vertex is odd.
+- `simple_graph.exists_ne_odd_degree_of_exists_odd_degree` is that the existence of an
+  odd-degree vertex implies the existence of another one.
+
+## Implementation notes
+
+We give a combinatorial proof by using the fact that the map from
+darts to vertices has fibers whose cardinalities are the degrees and
+that the map from darts to edges is 2-to-1.
+
+## Tags
+
+simple graphs, sums, degree-sum formula, handshaking lemma
+-/
+open finset
+
+open_locale big_operators
+
+namespace simple_graph
+universes u
+variables {V : Type u} (G : simple_graph V)
+
+/-- A dart is a directed edge, consisting of an ordered pair of adjacent vertices. -/
+@[ext, derive decidable_eq]
+structure dart :=
+(fst snd : V)
+(is_adj : G.adj fst snd)
+
+/-- There is an equivalence between darts and pairs of a vertex and an incident edge. -/
+@[simps]
+def dart_equiv_sigma : G.dart ≃ Σ v, G.neighbor_set v :=
+{ to_fun := λ d, ⟨d.fst, d.snd, d.is_adj⟩,
+  inv_fun := λ s, ⟨s.fst, s.snd, s.snd.property⟩,
+  left_inv := λ d, by ext; simp,
+  right_inv := λ s, by ext; simp }
+
+instance dart.fintype [fintype V] [decidable_rel G.adj] : fintype G.dart :=
+fintype.of_equiv _ G.dart_equiv_sigma.symm
+
+instance dart.inhabited [inhabited V] [inhabited (G.neighbor_set (default _))] :
+  inhabited G.dart := ⟨G.dart_equiv_sigma.symm ⟨default _, default _⟩⟩
+
+variables {G}
+
+/-- The edge associated to the dart. -/
+def dart.edge (d : G.dart) : sym2 V := ⟦(d.fst, d.snd)⟧
+
+@[simp] lemma dart.edge_mem (d : G.dart) : d.edge ∈ G.edge_set :=
+d.is_adj
+
+/-- Reverses the orientation of a dart. -/
+def dart.rev (d : G.dart) : G.dart :=
+⟨d.snd, d.fst, G.sym d.is_adj⟩
+
+@[simp] lemma dart.rev_edge (d : G.dart) : d.rev.edge = d.edge :=
+sym2.eq_swap
+
+@[simp] lemma dart.rev_rev (d : G.dart) : d.rev.rev = d :=
+dart.ext _ _ rfl rfl
+
+@[simp] lemma dart_rev_involutive : function.involutive (dart.rev : G.dart → G.dart) :=
```

Comment:
I wonder if all of these should be in the `dart` namespace. I don't really know what the convention is.

Reply:
I considered it suggesting this too, but I didn't see to many other namespaced `_involutive` declarations in a quick search.

Reply:
`list.reverse_involutive` is an easy counter-example. You probably didn't find any because many files would just use `namespace dart` and then put the definitions inside without any explicit `.`s

Reply:
Ah, OK. In that case do you think it should be `dart.rev.involutive` or `dart.rev_involutive`? I guess we're not likely to be able to use dot notation with the former.

(I was using the web docs search though which I thought includes full namespaces. However, that search [apparently has other issues](https://github.com/leanprover-community/doc-gen/issues/104).)

Reply:
`dart.rev_involutive` seems more consistent with other lemma names.

Reply:
```suggestion
@[simp] lemma dart.rev_involutive : function.involutive (dart.rev : G.dart → G.dart) :=
```